### PR TITLE
Allow specifying OS image in local env

### PR DESCRIPTION
**Pull request by pablochacin**
**What this PR does / why we need it**:
This PR modifies the specifications of the local cloud provider and the local shout in order to allow selecting the os image to be used. This change facilitates testing different OS images locally.

**Special notes for your reviewer**:

**Release note**:

```improvement operator
Allow specifying machine image name in local development environment to facilitate testing different images.
```

### CODE REVIEW

File: pkg/apis/garden/types.go [hunk 1]
```diff
@@ -320,6 +322,11 @@ type DNSProviderConstraint struct {
 	Name DNSProvider
 }
 
+// AlicloudMachineImage defines the machine image for Alicloud.
```
Comment (vpnachev) [inline pkg/apis/garden/types.go]:
```suggestion
// LocalMachineImage defines the machine image for Local provider.
```

File: pkg/apis/garden/types.go [hunk 2]
```diff
@@ -1011,6 +1018,11 @@ type Local struct {
 	Networks LocalNetworks
 	// Endpoint of the local service.
 	Endpoint string
+	// MachineImage holds information about the machine image to use for all workers.
+	// It will default to the first image stated in the referenced CloudProfile if no
+	// value has been provided.
+	// +optional
```
Comment (vpnachev) [inline pkg/apis/garden/types.go]: `optional` is not needed for the internal version

**Review by vpnachev — CHANGES_REQUESTED** [review]

File: pkg/apis/garden/v1beta1/types.go [hunk 3]
```diff
@@ -994,6 +1001,11 @@ type OpenStackWorker struct {
 type Local struct {
 	// Networks holds information about the Kubernetes and infrastructure networks.
 	Networks LocalNetworks `json:"networks"`
+	// MachineImage holds information about the machine image to use for all workers.
+	// It will default to the first image stated in the referenced CloudProfile if no
+	// value has been provided.
+	// +optional
+	MachineImage LocalMachineImage `json:"machineImage,omitempty"`
```
Comment (rfranzke) [inline pkg/apis/garden/v1beta1/types.go]:
```suggestion
	MachineImage *LocalMachineImage `json:"machineImage,omitempty"`
```

File: hack/dev-setup-local [hunk 4]
```diff
@@ -98,9 +98,6 @@ kubectl config use-context gardener-dev \
 # Different base64 implementations have different flags
 LOCAL_SEED_KUBECONFIG_B64=$(base64 ${LOCAL_SEED_KUBECONFIG} | tr -d '\r\n')
 
-sed -e "s/name: my-seed/name: local/g" example/25-controllerinstallation.yaml | \
```
Comment (rfranzke) [inline hack/dev-setup-local:101]: Why is that removed? If we need to remove it, can you please adapt the local dev. setup documentation accordingly?

**Review by rfranzke — CHANGES_REQUESTED** [review]
Who sets the `.spec.cloud.local.machineImage` field in the `Shoot` resource? I've not found any defaulting in the PR. If the user is expected to set it I suggest to update the example `Shoot` manifest for the local provider and include it there. Alternatively, if you don't want to "pin" it there please adapt the local dev. setup documentation.

File: hack/dev-setup-local (hunk 4 above)
Comment (pablochacin) [reply hack/dev-setup-local:101]: It is removed because now there's a separated step for extension installation. Therefore, this installation is not needed. No documentation change is required because this setup was executed implicitly (not documented).

**Comment by pablochacin** [comment]
> Who sets the `.spec.cloud.local.machineImage` field in the `Shoot` resource? I've not found any defaulting in the PR. If the user is expected to set it I suggest to update the example `Shoot` manifest for the local provider and include it there. Alternatively, if you don't want to "pin" it there please adapt the local dev. setup documentation.

I added the default to the `example/90-shoot-local.yaml`

**Comment by rfranzke** [comment]
Could you squash the commits and rebase? Also, please sign the CLA.

**Review by rfranzke — DISMISSED** [review]
/lgtm
Cannot be merged before the CLA has been signed.

File: example/30-cloudprofile-local.yaml [hunk 5]
```diff
@@ -12,3 +12,5 @@ spec:
     constraints:
       dnsProviders:
       - name: unmanaged
+      machineImages:
```
Comment (ialidzhikov) [inline example/30-cloudprofile-local.yaml:15]: You could add this logic into `30-cloudprofile.yaml.tpl`. After executing `./hack/generate-examples` your changes will be removed.
So apply your changes in templates files under `/hack/templates/resources` and run `./hack/generate-examples`.

File: example/90-shoot-local.yaml [hunk 6]
```diff
@@ -11,6 +11,8 @@ spec:
     secretBindingRef:
       name: core-local
     local:
+      machineImage:
```
Comment (ialidzhikov) [inline example/90-shoot-local.yaml]: Same.

**Review by ialidzhikov — CHANGES_REQUESTED** [review]

File: pkg/apis/garden/types.go [hunk 7]
```diff
@@ -1011,6 +1018,10 @@ type Local struct {
 	Networks LocalNetworks
 	// Endpoint of the local service.
 	Endpoint string
+	// MachineImage holds information about the machine image to use for all workers.
+	// It will default to the first image stated in the referenced CloudProfile if no
```
Comment (ialidzhikov) [inline pkg/apis/garden/types.go:1076]: I receive panic when I remove the `machineImage` part from the local shoot spec. You could remove the `defaulting` part from the doc or add the defaulting logic?

File: pkg/apis/garden/v1beta1/types.go [hunk 8]
```diff
@@ -994,6 +1001,11 @@ type OpenStackWorker struct {
 type Local struct {
 	// Networks holds information about the Kubernetes and infrastructure networks.
 	Networks LocalNetworks `json:"networks"`
+	// MachineImage holds information about the machine image to use for all workers.
+	// It will default to the first image stated in the referenced CloudProfile if no
```
Comment (ialidzhikov) [inline pkg/apis/garden/v1beta1/types.go:1059]: Same.

File: pkg/apis/garden/types.go (hunk 7 above)
Comment (pablochacin) [reply pkg/apis/garden/types.go:1076]: Fixed: default set at shoot validation plugin, as suggested by @rfranzke

File: pkg/apis/garden/v1beta1/types.go (hunk 8 above)
Comment (pablochacin) [reply pkg/apis/garden/v1beta1/types.go:1059]: Not sure what is expected here. Can you please elaborate?

File: hack/templates/resources/30-cloudprofile.yaml.tpl [hunk 9]
```diff
@@ -523,4 +523,6 @@ spec:<% caBundle=value("spec.caBundle", "") %>
       % else:
       - name: unmanaged
       % endif
+      machineImages:
+      - name: coreos
```
Comment (ialidzhikov) [inline hack/templates/resources/30-cloudprofile.yaml.tpl]:
```suggestion
      - name: coreos
```

File: hack/templates/resources/90-shoot.yaml.tpl [hunk 10]
```diff
@@ -201,6 +201,8 @@ spec:
     % endif
     % if cloud == "local":
     local:
+      machineImage:
+        name: coreos
```
Comment (ialidzhikov) [inline hack/templates/resources/90-shoot.yaml.tpl]:
```suggestion
        name: coreos
```

**Review by ialidzhikov — COMMENTED** [review]
/lgtm

Could you please also rebase?

**Comment by pablochacin** [comment]
> /lgtm
> 
> Could you please also rebase?

done

File: example/50-seed-alicloud.yaml [hunk 11]
```diff
@@ -4,7 +4,7 @@ apiVersion: garden.sapcloud.io/v1beta1
 kind: Seed
 metadata:
   name: alicloud
-  annotations: {persistentvolume.garden.sapcloud.io/minimumSize: 20Gi}
```
Comment (rfranzke) [inline example/50-seed-alicloud.yaml]: Why has this been changed? Can you please revert it?

**Review by rfranzke — CHANGES_REQUESTED** [review]

File: hack/templates/resources/90-shoot.yaml.tpl [hunk 12]
```diff
@@ -219,6 +219,8 @@ spec:
     % endif
     % if cloud == "local":
     local:
+      machineImage:
```
Comment (rfranzke) [inline hack/templates/resources/90-shoot.yaml.tpl]: Can we remove this now that we have defaulting in the admission?

**Comment by rfranzke** [comment]
Also, unfortunately, again rebase is needed (sth else was merged meanwhile).

File: hack/templates/resources/90-shoot.yaml.tpl (hunk 12 above)
Comment (pablochacin) [reply hack/templates/resources/90-shoot.yaml.tpl]: I think it would be interesting to have it there as an example, even if it is not strictly necessary, Maybe as a comment:
```
# specify the operating system image for local environment
# machineImage:
#   name: coreos
```

Comment (rfranzke) [reply hack/templates/resources/90-shoot.yaml.tpl]: OK, agreed. Let's make it a comment

**Review by rfranzke — APPROVED** [review]
/lgtm
